Copy constraints for splatting ops

liliatangxy opened the pull request:
The existing protocol for `PermuteBasicBlock` cannot handle operations like window-based splatting needed for proving Conv and pooling ops where an element can appear in multiple positions and when an index for one dimension of the permutation depends on indices of more than one dimension.

The basic block itself requires that the input and output last dimensions are the same, meaning that in order to perform a splat, we have to perform a shape op to make the dimensions the same. We should be able to do this with any basic block that can prove padding. However, `self.permutation` is based on the indices and shapes of the original padded input/output.  Right now it expands the last index of `self.permutation` such that the elements of the original are evenly spaced, but I can clean this up to just pad trailing zeros.

This draft PR implements the copy constraint check based on the Plonk paper Section 8 protocol, but without quotient polynomial batching.

arifiorino committed:
Repeater run

arifiorino committed:
encodeOutputs/prove/verify

arifiorino committed:
First 5 onnx ops

arifiorino committed:
Pass constants + pow + sqrt

liliatangxy committed:
plonk copy constraint

liliatangxy committed:
fix cargo run

arifiorino committed:
div + bug fixes

liliatangxy committed:
merge master

liliatangxy committed:
remove incorrect comment

arifiorino committed:
Starting reshape + transpose

ddkang reviewed (commented):
The setup seems complicated enough it should be a struct

ddkang commented on src/basic_block/copy_constraint.rs:
```diff
@@ -0,0 +1,397 @@
+
+// If is_input, idxs is [(flat_idx of the input index, 0)]. Otherwise,
+// idxs is [(flat_idx of the output, flat_idx of the permuted input idx)]
+fn construct_ssig_coeffs(idxs: &[(usize, usize)], partitions: &HashMap<usize, Vec<usize>>, is_input: bool) -> Vec<Fr> {
```
What is ssig?

liliatangxy replied:
I updated the comments for this function and `flat_index`. I can provide more detail if needed

ddkang replied:
I mean very literally, what is Sigma_j? Do you expect someone reading this file to know?

ddkang commented on src/basic_block/copy_constraint.rs:
```diff
@@ -0,0 +1,397 @@
+    let N = input_len;
+    let domain = GeneralEvaluationDomain::<Fr>::new(N).unwrap();
+
+    let beta = Fr::rand(rng);
```
Does this need to depend on anything? Can the prover cheat here?

liliatangxy replied:
At this point no new proof elements have been made since the previous basic block, so I think we don't need to make Fiat-Shamir adjustments here.

ddkang replied:
can you add a comment?

arifiorino committed:
Finished reshape + transpose

liliatangxy committed:
fiat-shamir

liliatangxy committed:
Revert "fiat-shamir"

This reverts commit 17911b1937817cc0301e21ea5f5d5644aa9d8acd.

liliatangxy committed:
fiat-shamir

Vimalakirti commented on src/basic_block/copy_constraint.rs:
```diff
@@ -0,0 +1,415 @@
+    assert!(Dim(self.permutation.shape()) == self.output_dim);
+
+    // The basic block inputs and outputs will be scaled to the same degree,
+    // meaning that the last dimensions will be the same.
```
What's the reason to scale them to the same degree?

Vimalakirti commented on src/basic_block/copy_constraint.rs:
```diff
@@ -0,0 +1,415 @@
+    return (proof_0, vec![L_i_x_2[0]], proof_2);
+  }
+
+  fn prove(
```
I feel this is exactly the same as the copy constraint described in Plonk. Is there anything I miss? (different from the paper?)

liliatangxy replied:
Yeah it is the same protocol

Vimalakirti reviewed (commented):
This PR implements the copy constraint (in Plonk).
Left some comments and also curious about Daniel's questions.
Please take a look at them, I'll review it carefully again later.

arifiorino commented on src/basic_block/copy_constraint.rs:
```diff
@@ -0,0 +1,415 @@
+#[derive(Debug)]
+pub struct CopyConstraintBasicBlock {
+  pub permutation: ArrayD<IxDyn>,
+  pub partitions: HashMap<usize, Vec<usize>>,
```
This variable can be moved to the `ProveVerifyCache`

liliatangxy replied:
It is unused so I removed it

arifiorino commented on src/tests.rs:
```diff
@@ -2,35 +2,41 @@ use crate::basic_block::*;
 use rand::{rngs::StdRng, SeedableRng};
 use std::collections::HashMap;
 
+use self::copy_constraint::CopyConstraintBasicBlock;
```
`CopyConstraintBasicBlock` should be added to the file basic_block.rs and then we can delete this line because all the basic blocks are imported on line 1 of this file.

arifiorino commented on src/basic_block/copy_constraint.rs:
```diff
@@ -0,0 +1,415 @@
+
+      if i[i.ndim() - 1] % (N / last_outp_dim) == 0 {
+        let mut inp_idx = self.permutation[&outp_idx].clone();
+        inp_idx[self.input_dim.ndim() - 1] = inp_idx[self.input_dim.ndim() - 1] * N / last_inp_dim;
```
These two lines are unnecessary. You define a new variable `inp_idx`, then modify it, and then never use it.

arifiorino commented on src/basic_block/copy_constraint.rs:
```diff
@@ -0,0 +1,415 @@
+    // intput and output, so we need to adjust the input and output indices
+    // to correspond to permutation and fill in the output values that are
+    // not part of the actual permutation
+    let last_inp_dim = self.input_dim[self.input_dim.ndim() - 1];
```
This can be replaced with `self.input_dim.last().unwrap()`.

liliatangxy replied:
Doesn't seem to work for `IxDyn`

arifiorino commented on src/basic_block/copy_constraint.rs:
```diff
@@ -0,0 +1,415 @@
+    // not part of the actual permutation
+    let last_inp_dim = self.input_dim[self.input_dim.ndim() - 1];
+    let last_outp_dim = self.output_dim[self.output_dim.ndim() - 1];
+    let N = max(self.input_dim[self.input_dim.ndim() - 1], self.output_dim[self.output_dim.ndim() - 1]);
```
`tests.rs` doesn't have a test case for if these two values are different (it never tests for the padding). Can you add a test case for this?

arifiorino replied:
Also, it seems that you pad the output if the input is smaller than the output. But what about the other case? What if the output is smaller than the input?

arifiorino reviewed (changes requested):
Can you write a Convolution `Layer`? Then it would be easy to plug it into the onnx framework.

arifiorino committed:
mul layer

liliatangxy committed:
fix padding

liliatangxy commented:
Fixed some of the padding stuff and added tests for it.
It turns out that with the existing padding method we do not need to pad the inputs and outputs, we just need to pad the setup and permutation indices, so we no longer need PermuteBasicBlock before or after for dimension changes.

liliatangxy committed:
update comments

liliatangxy committed:
update comments

liliatangxy committed:
merge distilbert

liliatangxy committed:
merge distilbert

liliatangxy committed:
spacing fix
